src/uct/tcp/tcp_ep.c

```diff
@@ -1589,18 +1589,73 @@ static void uct_tcp_ep_post_put_ack(uct_tcp_ep_t *ep)
     ep->flags &= ~UCT_TCP_EP_FLAG_PUT_RX_SENDING_ACK;
 }
 
+/* payload and count have different meanings depending on is_short_iov. If
+ * is_short_iov is 0, then payload is a contiguous buffer and count is a size of the
+ * buffer. Otherwise, payload points to the beginning of an array of uct_iov_t, and
+ * count is the size of the array. */
+static inline ucs_status_t
+uct_tcp_ep_am_short_sendv(uct_tcp_ep_t *ep, uct_tcp_iface_t *iface,
+                          uct_tcp_am_hdr_t *hdr, uint64_t header,
+                          const void *payload, size_t count, struct iovec *iov,
+                          int is_short_iov)
+{
+    ucs_iov_iter_t uct_iov_iter;
+    size_t iov_cnt;
+    ucs_status_t status;
+    size_t offset;
+
+    iov[0].iov_base = hdr;
+    iov[0].iov_len  = sizeof(*hdr);
+
+    if (is_short_iov) {
```

Comment:
it seems somewhat confusing that behavior is so different based on is_short_iov.. maybe pass already full "iov" list to this function, and fill it in caller functions? IMO the duplication of 1607-1608 lines is not so critical

Reply:
maybe it's worth to allocate IOVs thru `ucs_alloca()` here instead of in the caller's function? @yosefe wdyt (or `count` is a confusing point here)?
i.e.:
```
struct iovec *iov = ucs_alloca(is_short_iov ?
                               ((iov_cnt + 1) * sizeof(*iov)) :
                               UCT_TCP_EP_AM_SHORTV_IOV_COUNT);
```
it will decrease duplication for allocation IOVs

Reply:
IMO, count and is_short_iov is the confusing point here

Reply:
> IMO, count and is_short_iov is the confusing point here

I see, thanks 👍 
but `is_short_iov` still should be passed to `uct_tcp_ep_am_sendv()` as the second argument, so, we will keep it (probably, should be renamed to `shortv_send` - means normal AM short, but using IOV send)

Reply:
> but `is_short_iov` still should be passed to `uct_tcp_ep_am_sendv()` as the second argument, so, we will keep it (probably, should be renamed to `shortv_send` - means normal AM short, but using IOV send)

looks like it's only for tracing, maybe maybe `uct_tcp_ep_am_sendv_get_trace_payload` can use iov count to determine if can return pointer to header or need to copy

Reply:
> > but `is_short_iov` still should be passed to `uct_tcp_ep_am_sendv()` as the second argument, so, we will keep it (probably, should be renamed to `shortv_send` - means normal AM short, but using IOV send)
> 
> looks like it's only for tracing, maybe maybe `uct_tcp_ep_am_sendv_get_trace_payload` can use iov count to determine if can return pointer to header or need to copy

yes, it is used only for tracing. but if we will rely on iov count, it could lead to wrong behavior of trace function - when it will return an address of `iov[1]` instead of copying to buffer for AM Short IOV with `iov_count = 2` case

Reply:
maybe we can always copy whole iov to some temp buffer and use it for tracing?

Reply:
> maybe we can always copy whole iov to some temp buffer and use it for tracing?

I see, makes sense